deploy/AllocatorDeploy.sol

```diff
@@ -0,0 +1,97 @@
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity ^0.8.16;
+
+import { ScriptTools } from "dss-test/ScriptTools.sol";
+
+import { AllocatorOracle }      from "src/AllocatorOracle.sol";
+import { AllocatorRoles }       from "src/AllocatorRoles.sol";
+import { AllocatorRegistry }    from "src/AllocatorRegistry.sol";
+import { AllocatorBuffer }      from "src/AllocatorBuffer.sol";
+import { AllocatorVault }       from "src/AllocatorVault.sol";
+import { Swapper }              from "src/funnels/Swapper.sol";
+import { DepositorUniV3 }       from "src/funnels/DepositorUniV3.sol";
+import { StableSwapper }        from "src/funnels/automation/StableSwapper.sol";
+import { StableDepositorUniV3 } from "src/funnels/automation/StableDepositorUniV3.sol";
+import { ConduitMover }         from "src/funnels/automation/ConduitMover.sol";
+
+import { AllocatorSharedInstance, AllocatorNetworkInstance } from "./AllocatorInstances.sol";
+
+interface NstJoinLike {
+    function vat() external view returns (address);
+}
+
+library AllocatorDeploy {
+
+    // Note: owner is assumed to be the pause proxy
+    function deployShared(
+        address deployer,
+        address owner
+    ) internal returns (AllocatorSharedInstance memory sharedInstance) {
+        address _oracle = address(new AllocatorOracle());
+
+        address _roles  = address(new AllocatorRoles());
+        ScriptTools.switchOwner(_roles, deployer, owner);
+
+        address _registry = address(new AllocatorRegistry());
+        ScriptTools.switchOwner(_registry, deployer, owner);
+
+        sharedInstance.oracle   = _oracle;
+        sharedInstance.roles    = _roles;
+        sharedInstance.registry = _registry;
+    }
+
+    // Note: owner is assumed to be the pause proxy, allocatorProxy will receive ownerships on init
+    function deployNetwork(
+        address deployer,
+        address owner,
+        address roles,
+        bytes32 ilk,
+        address nstJoin,
+        address uniV3Factory
+    ) internal returns (AllocatorNetworkInstance memory networkInstance) {
+        address _buffer = address(new AllocatorBuffer());
+        ScriptTools.switchOwner(_buffer, deployer, owner);
+        networkInstance.buffer = _buffer;
+
+        address _vault  = address(new AllocatorVault(roles, _buffer, NstJoinLike(nstJoin).vat(), ilk, nstJoin));
```

Comment:
Actually the `vat` check in Vault is a bit dumb if we are passing the values this way. Not necessarily asking to change anything yet but just wondering.

Reply:
Yes I know.
It makes sense so remove the vat parameter from the vault but I didn't want to touch that at this PR.
Can still do it though.

Reply:
Hmm yeah let's leave it as it is, and then we can do a separate PR to modify the vault and the deployer together.